### PR TITLE
init changes for replaceable peers in k bucket

### LINKED ISSUE #110: Displace oldest peer in kademlia's peer routing backet at overflow
Displace oldest peer in kademlia's peer routing table backet at overflow. Right now it returns with error 'no capacity'.

### PR DESCRIPTION
**Pull request by vvarma**
fixes #110 
implemented similar to https://github.com/libp2p/go-libp2p-kbucket/blob/master/table.go

Please let me know if the above PR is in alignment with what's expected from this module. Also looking forward to comments about coding guidelines and other semantic gotchas.

### CODE REVIEW

File: test/libp2p/protocol/kademlia/node_id_test.cpp [hunk 1]
```diff
@@ -31,20 +34,20 @@ bool is_distance_less(Hash256 a, Hash256 b) {
...
 
-bool is_xor_distance_sorted(const PeerId &local, std::vector<PeerId> &peers) {
+bool is_xor_distance_sorted(const PeerId &local, std::vector<BucketPeerInfo> &peers) {
   NodeId nlocal(local);
 
   auto begin = peers.begin();
   auto end = peers.end();
   while (begin != end) {
-    NodeId nremote1(*begin);
+    NodeId nremote1((*begin).node_id);
```
Comment (xDimon) [inline test/libp2p/protocol/kademlia/node_id_test.cpp:43]:
```suggestion
    const auto& nremote1 = (*begin).node_id;
```

File: test/libp2p/protocol/kademlia/node_id_test.cpp [hunk 2]
```diff
@@ -31,20 +34,20 @@ bool is_distance_less(Hash256 a, Hash256 b) {
...
     auto distance1 = nremote1.distance(nlocal);
 
     if (++begin == end) {
       continue;
     }
 
-    NodeId nremote2(*begin);
+    NodeId nremote2((*begin).node_id);
```
Comment (xDimon) [inline test/libp2p/protocol/kademlia/node_id_test.cpp:50]:
```suggestion
    const auto& nremote2 = (*begin).node_id;
```

File: src/protocol/kademlia/impl/peer_routing_table_impl.cpp [hunk 3]
```diff
@@ -88,25 +88,24 @@ namespace libp2p::protocol::kademlia {
...
     size_t cpl = nodeId.commonPrefixLen(local_);
 
     auto bucketId = getBucketId(buckets_, cpl);
     auto &bucket = buckets_.at(bucketId);
 
     // Trying to find and move to front
-    auto it = std::remove(bucket.rbegin(), bucket.rend(), pid);
-    if (it != bucket.rend()) {
-      *it = pid;
+    if (bucket.contains(pid)) {
```
Comment (xDimon) [inline src/protocol/kademlia/impl/peer_routing_table_impl.cpp:103]: I'm not sure if it is enough. I guess peer ids in backet must be ordered by time of add/update.

File: test/libp2p/protocol/kademlia/node_id_test.cpp [hunk 4]
```diff
@@ -1,9 +1,11 @@
+#include <include/libp2p/protocol/kademlia/impl/peer_routing_table_impl.hpp>
```
Comment (xDimon) [inline test/libp2p/protocol/kademlia/node_id_test.cpp:1]:
```suggestion
```

File: test/libp2p/protocol/kademlia/peer_routing_table_test.cpp [hunk 5]
```diff
@@ -54,9 +54,8 @@ TEST_F(PeerRoutingTableTest, BusWorks) {
 
   std::unordered_set<PeerId> peerset;
 
-  auto addHandle = addCh.subscribe([&](const PeerId &pid) {
-    peerset.insert(pid);
-  });
+  auto addHandle =
+      addCh.subscribe([&](const PeerId &pid) { peerset.insert(pid); });
```
Comment (xDimon) [inline test/libp2p/protocol/kademlia/peer_routing_table_test.cpp:58]: Please, use clang-format tool to be in according with using code-style. We are using Google code style, defined in file [.clang-format](https://github.com/libp2p/cpp-libp2p/blob/master/.clang-format)

File: test/libp2p/protocol/kademlia/peer_routing_table_test.cpp [hunk 6]
```diff
@@ -120,11 +174,39 @@ TEST_F(PeerRoutingTableTest, EldestPreferred) {
...
     } else {
       // Remained are rejected
       ASSERT_OUTCOME_ERROR(
-          table_->update(peers[i]),
+          table_->update(peers[i], true),
+          PeerRoutingTableImpl::Error::PEER_REJECTED_NO_CAPACITY);
+    }
+  }
+}
+TEST_F(PeerRoutingTableTest, EldestPrefferedIfPermanent) {
```
Comment (xDimon) [inline test/libp2p/protocol/kademlia/peer_routing_table_test.cpp:186]:
```suggestion
}

TEST_F(PeerRoutingTableTest, EldestPrefferedIfPermanent) {
```

File: test/libp2p/protocol/kademlia/peer_routing_table_test.cpp [hunk 7]
```diff
@@ -101,7 +100,62 @@ TEST_F(PeerRoutingTableTest, FindMultiple) {
...
+
+  // re-adding an existent peer should return false regardless of is permanent
+  auto updateVal = table_->update(peers[2], true);
+  ASSERT_OUTCOME_SUCCESS_TRY(updateVal);
+  ASSERT_FALSE(updateVal.value());
+  updateVal = table_->update(peers[2], false);
+  ASSERT_OUTCOME_SUCCESS_TRY(updateVal);
+  ASSERT_FALSE(updateVal.value());
+}
+TEST_F(PeerRoutingTableTest, EldestRecycledIfNotPermanent) {
```
Comment (xDimon) [inline test/libp2p/protocol/kademlia/peer_routing_table_test.cpp:158]:
```suggestion
}

TEST_F(PeerRoutingTableTest, EldestRecycledIfNotPermanent) {
```

File: src/protocol/kademlia/impl/peer_routing_table_impl.cpp [hunk 8]
```diff
@@ -119,22 +118,40 @@ namespace libp2p::protocol::kademlia {
...
-        bid = buckets_.size() - 1;
+      auto resizedBucketId = getBucketId(buckets_, cpl);
+      auto &resizedBucket = buckets_.at(resizedBucketId);
+      if (resizedBucket.size() < config_.maxBucketSize) {
+        resizedBucket.emplace_front(pid, is_permanent);
+        bus_->getChannel<events::PeerAddedChannel>().publish(pid);
+        return true;
       }
+      auto replaceablePeerIt =
+          std::find_if(resizedBucket.begin(), resizedBucket.end(),
```
Comment (xDimon) [inline src/protocol/kademlia/impl/peer_routing_table_impl.cpp:129]: Perhaps is better to use reversive search. Eldest peer ids are expected it tail.
```suggestion
          std::find_if(resizedBucket.rbegin(), resizedBucket.rend(),
```

File: src/protocol/kademlia/impl/peer_routing_table_impl.cpp [hunk 9]
```diff
@@ -119,22 +118,40 @@ namespace libp2p::protocol::kademlia {
...
+      auto removedPeer = (*replaceablePeerIt).peer_id;
+      bus_->getChannel<events::PeerRemovedChannel>().publish(removedPeer);
+      resizedBucket.erase(replaceablePeerIt);
+      resizedBucket.emplace_front(pid, !is_permanent);
       bus_->getChannel<events::PeerAddedChannel>().publish(pid);
-      return false;
+      return true;
     }
+    auto replaceablePeerIt =
+        std::find_if(bucket.begin(), bucket.end(),
```
Comment (xDimon) [inline src/protocol/kademlia/impl/peer_routing_table_impl.cpp:143]: The same

File: src/protocol/kademlia/impl/peer_routing_table_impl.cpp (hunk 3 above)
Comment (vvarma) [reply src/protocol/kademlia/impl/peer_routing_table_impl.cpp:103]: Thanks @xDimon for pointing this out.
From the go-libp2p implementation 
`// If the logical bucket to which the peer belongs is full and it's not the last bucket, we try to replace an existing peer
// whose LastSuccessfulOutboundQuery is above the maximum allowed threshold in that bucket with the new peer.
// If no such peer exists in that bucket, we do NOT add the peer to the Routing Table and return error "ErrPeerRejectedNoCapacity".`
Shall I try to implement it in a similar fashion?

**Comment by vvarma** [comment]
@xDimon thanks for the review! I have updated the PR with the changes requested and have tried to follow the specs as closely as possible. I have also formatted the changes as requested using the clang-format (but the line you mentioned, is still being formatted in the same way, will check and get back here).

File: src/protocol/kademlia/impl/peer_routing_table_impl.cpp (hunk 3 above)
Comment (xDimon) [reply src/protocol/kademlia/impl/peer_routing_table_impl.cpp]: It sounds reasonable. Don't stop, please! 👍

File: include/libp2p/protocol/kademlia/impl/peer_routing_table.hpp [hunk 10]
```diff
@@ -33,7 +33,9 @@ namespace libp2p::protocol::kademlia {
      * Adds/moves the given @param peer to the front of its respective bucket
      * @returns true/false if it was newadded peer or not
      */
-    virtual outcome::result<bool> update(const peer::PeerId &peer, bool is_permanent) = 0;
+    virtual outcome::result<bool> update(const peer::PeerId &peer,
+                                         bool is_permanent,
+                                         bool is_connected=false) = 0;
```
Comment (xDimon) [inline include/libp2p/protocol/kademlia/impl/peer_routing_table.hpp:38]: Seems clang-format did not apply for this file

File: src/protocol/kademlia/impl/peer_routing_table_impl.cpp [hunk 11]
```diff
@@ -126,29 +136,31 @@ namespace libp2p::protocol::kademlia {
...
                        [](const auto &bpi) { return bpi.is_replaceable; });
 
-      if (replaceablePeerIt == resizedBucket.end()) {
+      if (replaceablePeerIt == resizedBucket.rend()) {
         return Error::PEER_REJECTED_NO_CAPACITY;
       }
       auto removedPeer = (*replaceablePeerIt).peer_id;
       bus_->getChannel<events::PeerRemovedChannel>().publish(removedPeer);
-      resizedBucket.erase(replaceablePeerIt);
+      std::advance(replaceablePeerIt, 1);
```
Comment (xDimon) [inline src/protocol/kademlia/impl/peer_routing_table_impl.cpp:147]: I guess, `++replaceablePeerIt` is more human readable. `std::advance` is more situable for change for greater than one step.

**Review by kamilsa — APPROVED** [review]

**Review by xDimon — APPROVED** [review]

**Comment by vvarma** [comment]
thanks @kamilsa @xDimon !

**Comment by kamilsa** [comment]
> thanks @kamilsa @xDimon !

@vvarma thanks for your contribution!

**Pull request merged**